volumes/miovision/sql/table/create_anomalous_ranges.sql

```diff
@@ -30,6 +31,8 @@ CREATE TABLE miovision_api.anomalous_ranges (
     intersection_uid smallint REFERENCES miovision_api.intersections (intersection_uid),
     classification_uid smallint REFERENCES miovision_api.classifications (classification_uid),
     time_range tsrange NOT NULL,
+    range_start timestamp NOT NULL,
+    range_end timestamp,
```

Comment:
Aren't these redundant to the `time_range` field? You could get them with 

```sql
lower(time_range),
upper(time_range)
```

https://www.postgresql.org/docs/current/functions-range.html

Reply:
What was the problem with indexing?

Reply:
I created a small test case and in-line with Raph's suggestion, the explicit range_start and range_end was able to use the volumes datetime_bin index and ran about 50% faster. 
Explain Plan using `time_range @> datetime_bin`:
![image](https://github.com/CityofToronto/bdit_data-sources/assets/80077912/aba3ff37-c5ec-415e-a58d-acc03615455e)
vs
Explain plan using `datetime_bin >= LOWER(time_range) AND datetime_bin < UPPER(time_range)` (and similar for `datetime_bin >= range_start AND datetime_bin < range_end`:
![image](https://github.com/CityofToronto/bdit_data-sources/assets/80077912/52111dcf-67dd-4c40-a483-bb276ae84aaf)

Reply:
Sorry, I'm not sure I'm totally following. @gabrielwol are you saying that creating an index on certain fields on the `anomalous_ranges` table speeds up the join with one or more of the volumes tables? 

I don't know why you would ever do `datetime_bin >= LOWER(time_range) AND datetime_bin < UPPER(time_range)` rather than `time_range @> datetime_bin`. It's not surprising that's less efficient, right?

Reply:
@Nate-Wessel other way around: `datetime_bin >= LOWER(time_range) AND datetime_bin < UPPER(time_range)` was 50% faster than `time_range @> datetime_bin`.
The above screenshots were showing that the ts_range method was NOT able to use the index on the datetime_bin column of volumes, while the explicit method was.

Reply:
Oh, huh. Interesting! I'm surprised Postgres hasn't optimized this yet for these built-in types and operators. I'd expect the planner to be smarter than that. 

If that works efficiently with the `lower` and `upper` functions in the query, can we remove the duplicate fields on the table? Alternatively we could remove the tsrange field, though there are already a number of existing things that rely on that - they would have to be refactored. 

I'd just really like to avoid having three columns all saying the same thing in slightly different ways.

Reply:
>Oh, huh. Interesting! I'm surprised Postgres hasn't optimized this yet for these built-in types and operators. I'd expect the planner to be smarter than that.

I'm sure they have their reasons, but yeah I was also surprised by this. There are a lot of optimizations that seem intuitive for timestamps and yet... they don't exist.

I agree we shouldn't have duplicative columns (though we can ensure they are equivalent with triggers, but that's a real belt & suspender situation). My preference would be to use the timestamp columns going forward, so people don't get tripped up with the ranges and get sad when their queries take way longer than they expect. If you've got higher priorities than refactoring we could punt deleting the `tsrange` column until you can take that on?

Reply:
There's only two dependents in bigdata: `gwolofs.hourly_vehicle_volumes`, `monitoring.vi_daily_totals_miovision`
and one reference in bigdata (in this PR) that we'll have to update as far as I can tell.

Reply:
I'm ok with that. Sad that the thing that seems like it should be the clear solution isn't great though :-/ Oh well. 

May I suggest an interim check constraint rather than a trigger? I've found triggers to be fairly opaque. Something like:

```sql
CHECK(range_start = lower(time_range)),
CHECK(range_end = upper(time_range))
```
I'm not 100% sure how that works with `[)` vs `[]` etc but it shouldn't be too hard to figure out.

Reply:
Micro lightning talk: you can use `lower_inc(time_range), upper_inc(time_range)` to check if bounds are inclusive/exclusive! 
https://www.postgresql.org/docs/current/rangetypes.html#RANGETYPES-INCLUSIVITY